### ISSUE TITLE
NormFace/prototxt/scaled_cosine_softmax.prototxt 

### ISSUE THREAD

**Issue by ZHAIXINGZHAIYUE** [happynear/NormFace#4]
In the layer:id_weight_ip ,there is no bottom is found? why?

**Comment by ZHAIXINGZHAIYUE** [comment]
maybe ,i get your idea .when finetune a mdel, the model's last innerproduct layer( used to classfication, before softmaxLoss layer)  is adopted, we need to lear a new w for classification. right？

**Comment by happynear** [comment]
Correct, you may also try another strategy, adding a parameter "normalize: true" in the inner_product_param. This parameter performs a "hard normalization"(force to normalize the weights when forward). The performance is similar with the persent solution.

**Comment by happynear** [comment]
By this strategy, you can directly fine-tune the innerproduct layer without re-initialization. However, my implementation is not elegent. A friend told me to refact it as the bias term in scale layer. Maybe I will do this in next month.

**Comment by ZHAIXINGZHAIYUE** [comment]
Thanks. I am finetuning the wu's model.is not so good.
The accuracy on the LFW : from 0.989 to 0.9898.
the batch_size is 32 and training with one gpu
`base_lr: 0.001
lr_policy: "multistep"
gamma: 0.1
iter_size: 2
stepvalue: 5000
stepvalue: 10000
stepvalue: 150000
max_iter: 20000
momentum: 0.9
weight_decay: 0.0005
`

would you mind to give me some advices to reproduce your result? 
Thank you very much.

**Comment by happynear** [comment]
With Wu's model, we usually need more iterations. For example, I fine-tuned the model by 10,000 iterations with lr of 0.001, and another 16,000 iterations with lr of 0.0001. My batchsize is 32*4 (4 gpus).

With Wen's ResNet model, I only trained for 6,000 iterations with lr of 0.001. Batchsize is 64*4.

**Comment by happynear** [comment]
BTW, which dataset are you using for training?

**Comment by ZHAIXINGZHAIYUE** [comment]
i use the casia dataset(webface) to train .
another , i use the wu's model to test on the lfw with the mirror trick ,it doesn't work.
i use my caffe to  extract the the feature of original img and mirrored img respectively, then doing  max operation eletwise.

**Comment by happynear** [comment]
Wu's model (MaxOut) can achieve 98.9% using webface as training dataset? 

For the mirror trick, is it not better than single original image?

**Comment by ZHAIXINGZHAIYUE** [comment]
I means that i test the original wu's model: lightened_cnn_B ,i found that the mirror trick dosen't work .
Maybe , the reason why pca dosen't work on wu's modle  is that : wu use the random crop in data layer.

**Comment by happynear** [comment]
The results of mirror trick on lightened_cnn_B is here https://github.com/happynear/NormFace/blob/master/MirrorFace.md .

| PCA? | Front only  | Concatenate | Element-wise SUM | Element-wise MAX |
| ---- |:-----------:|:-----------:|:----------------:|:----------------:|
| No   | 98.10%      |98.35%       |98.42%            |98.35%            |
| YES  | 98.41%      |98.63%       |98.61%            |98.63%            |

It may not boost the performance compared with concatenate. But I never see cases that the mirror trick doesn't work compared with only one side of face image.

**Comment by happynear** [comment]
For the PCA, I still can't get the point why random crop can improve the performance of PCA. When fine-tuning Wu's model, I also used the random crop strategy.

I think the effect of PCA is some kind of domain adaption. However, the source of webface and lfw is similar. PCA shouldn't be so effective when the training set is webface.

**Comment by ZHAIXINGZHAIYUE** [comment]
Do you have use the dropout layer to fine-tune the wu's modle?

**Comment by ZHAIXINGZHAIYUE** [comment]
here is my result:
And with [Wu's Light CNN B](https://github.com/AlfredXiangWu/face_verification_experiment):

| PCA? | Front only  | Element-wise MAX |
| ---- |:----------------:|:----------------:|
| No   | 98.15%      |98.17%            |
| YES  | 98.42%      |98.33%            |

the aligned LFW dataset from here :链接: http://pan.baidu.com/s/1jIOAfrw 密码: bcm5 .(https://github.com/AlfredXiangWu/face_verification_experiment/issues/91)

**Comment by ZHAIXINGZHAIYUE** [comment]
Would you mind give me your aligned lfw dataset? Thank you.

**Comment by happynear** [comment]
I tried to use dropout, but using dropout have no help on my loss function. 

It seems that our results on front only is similar, but have big difference in eltwise max. I think there is something wrong with your codes... I think you should make sure that the images are flipped correctly. Maybe you can use my deploy prototxt https://github.com/happynear/NormFace/blob/master/prototxt/example_of_mirror_face.prototxt .

**Comment by ZHAIXINGZHAIYUE** [comment]
The mirror trick still doesn't work，although I use your caffe and your deploy prototxt .Maybe ,the key is the method to align the face image.

**Comment by happynear** [comment]
The aligned LFW images: https://drive.google.com/open?id=0B0OhXbSTAU1HTU91T1I0bU5DV2c

**Comment by ZHAIXINGZHAIYUE** [comment]
Yes,It does work.Maybe , there is something wrong in my code.Thank you.

**Comment by BrettLL** [comment]
@happynear  in your inner_product layer, you norm w in forward, but in backward why do you ignore the norm ??

**Comment by happynear** [comment]
@BrettLL, I directly modified the weight blob during forward. A better solution is written in the `scaled_cosine_softmax.prototxt`, using a parameter layer and a normalize layer to do the weight normalization.

**Comment by BrettLL** [comment]
@happynear  when I finetune the model , there is a big difference whether I use shuffle or not in my datalayer. This surprises me, have you met this problem??

**Comment by happynear** [comment]
@BrettLL, this is a well-known problem. If you do not shuffle the data, all the data in one batch may have the same label. All the features are encouraged to go to the same direction. This would cause the network learn nothing in a batch. Please keep in mind always shuffle your data during training.

BTW, when you are asking a new question, please open a new issue. It will be easier for other people who have the same question to search.Thx.

**Comment by BrettLL** [comment]
@happynear sorry, I didn't tell the problem clear. When I use shuffle ture, I got a worse result than set it to false. So, this amazes me. I will open a new issue if I have a new question~

**Comment by happynear** [comment]
...
I don't know how to explain this.. DL is magic.
